src/weighted_list.h

```diff
@@ -0,0 +1,107 @@
+#ifndef WEIGHTED_LIST_H
+#define WEIGHTED_LIST_H
+
+#include "rng.h"
+#include <vector>
+
+template <typename T> struct weighted_obect {
+    T obj;
+    int weight;
+};
+
+template <typename T> struct weighted_list {
+    weighted_list() : total_weight(0) { };
+
+    /**
+     * This will add a new object to the weighted list.
+     * @param obj The object that will be added to the list.
+     * @param weight The weight of the object.
+     */
+    void add_item(T obj, int weight) {
```

Comment:
Use a const reference as parameter. This does currently copy the object. It's copied yet another time when the `new_weighted` object is created below, consider using vectors `emplace_back` method instead.

Same for the other add function.